File: src/main/java/com/jetbrains/pluginverifier/verifiers/instruction/TypeInstructionVerifier.java
package com.jetbrains.pluginverifier.verifiers.instruction;

import com.jetbrains.pluginverifier.ErrorRegister;
import com.jetbrains.pluginverifier.resolvers.Resolver;
import com.jetbrains.pluginverifier.verifiers.util.VerifierUtil;
import org.objectweb.asm.tree.AbstractInsnNode;
import org.objectweb.asm.tree.ClassNode;
import org.objectweb.asm.tree.MethodNode;
import org.objectweb.asm.tree.TypeInsnNode;

/**
 * @author Dennis.Ushakov
 */
public class TypeInstructionVerifier implements InstructionVerifier {
  public void verify(final ClassNode clazz, final MethodNode method, final AbstractInsnNode instr, final Resolver resolver, final ErrorRegister register) {
    if (!(instr instanceof TypeInsnNode)) return;
    final String className = ((TypeInsnNode)instr).desc;
    if(className == null || VerifierUtil.classExists(resolver, className)) return;
    register.registerError(clazz.name + "." + method.name, "class " + className + " not found");
  }
}

File: intellij-plugin-verifier/verifier-core/src/main/java/com/jetbrains/pluginverifier/verifiers/instruction/TypeInstructionVerifier.kt
/*
 * Copyright 2000-2020 JetBrains s.r.o. and other contributors. Use of this source code is governed by the Apache 2.0 license that can be found in the LICENSE file.
 */

package com.jetbrains.pluginverifier.verifiers.instruction

import com.jetbrains.pluginverifier.results.problems.AbstractClassInstantiationProblem
import com.jetbrains.pluginverifier.results.problems.InterfaceInstantiationProblem
import com.jetbrains.pluginverifier.verifiers.VerificationContext
import com.jetbrains.pluginverifier.verifiers.extractClassNameFromDescriptor
import com.jetbrains.pluginverifier.verifiers.resolution.Method
import com.jetbrains.pluginverifier.verifiers.resolution.resolveClassChecked
import org.objectweb.asm.Opcodes
import org.objectweb.asm.tree.AbstractInsnNode
import org.objectweb.asm.tree.TypeInsnNode

/**
 * Processing of `new`, `anewarray`, `checkcast`, `instanceof` instructions.
 */
class TypeInstructionVerifier : InstructionVerifier {
  override fun verify(method: Method, instructionNode: AbstractInsnNode, context: VerificationContext) {
    if (instructionNode !is TypeInsnNode) return

    val className = instructionNode.desc.extractClassNameFromDescriptor() ?: return

    val typeClassFile = context.classResolver.resolveClassChecked(className, method, context) ?: return

    if (instructionNode.opcode == Opcodes.NEW) {
      if (typeClassFile.isInterface) {
        context.problemRegistrar.registerProblem(InterfaceInstantiationProblem(typeClassFile.location, method.location))
      } else if (typeClassFile.isAbstract) {
        context.problemRegistrar.registerProblem(AbstractClassInstantiationProblem(typeClassFile.location, method.location))
      }
    }

  }

}
